zebra-state/src/in_memory.rs

```diff
@@ -0,0 +1,119 @@
+use super::{Request, Response};
+use futures::prelude::*;
+use std::{
+    collections::{BTreeMap, HashMap},
+    error::Error,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+#[derive(Default)]
+struct ZebraState {
+    by_hash: HashMap<BlockHeaderHash, Arc<Block>>,
+    by_height: BTreeMap<BlockHeight, Arc<Block>>,
+}
+
+impl ZebraState {
+    fn insert(&mut self, block: impl Into<Arc<Block>>) {
+        let block = block.into();
+        let hash = block.as_ref().into();
+        let height = block.coinbase_height().unwrap();
+
+        assert!(
+            self.by_hash.insert(hash, block.clone()).is_none(),
+            "blocks shouldn't have the same hash"
+        );
+        assert!(
+            self.by_height.insert(height, block).is_none(),
+            "blocks with the same height are currently unsupported"
+        );
+    }
+
+    fn get(&mut self, query: impl Into<BlockQuery>) -> Option<Arc<Block>> {
+        match query.into() {
+            BlockQuery::ByHash(hash) => self.by_hash.get(&hash),
+            BlockQuery::ByHeight(height) => self.by_height.get(&height),
+        }
+        .cloned()
+    }
+
+    fn tip(&self) -> Option<Arc<Block>> {
+        self.by_height
+            .iter()
+            .next_back()
+            .map(|(_key, value)| value)
+            .cloned()
+    }
+}
+
+impl Service<Request> for ZebraState {
+    type Response = Response;
+    type Error = Box<dyn Error + Send + Sync + 'static>;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        match req {
+            Request::AddBlock { block } => {
+                self.insert(block);
+                async { Ok(Response::Added) }.boxed()
+            }
+            Request::GetBlock { hash } => {
+                let result = self
+                    .get(hash)
+                    .map(|block| Response::Block { block })
+                    .ok_or_else(|| "block could not be found".into());
+
+                async move { result }.boxed()
+            }
+            Request::GetTip => {
```

Comment:
I would match naming between the request type and the method name, so `tip()` would be `get_tip()`, also to reflect the type of operation on the datastore (CRUD, R-> 'get', etc)